Update readme (#256)
GeoscienceAustralia/dea-sandbox: README.md

Builds are run automatically out of this repository. Basic workflow is:

- Create a branch and implement your changes.

> - The Docker image is built on the branch and as part of the pull request.

Reviewer: I'm a bit puzzled about the 'as part of the pull request' bit. 
```suggestion
- A Docker image will be built when changes are pushed to the branch.
```

Author: There will be other actions triggered with PR, so docker image build is only part of the whole workflow. Similar to the next sentence, docker image build will be run on both PR and push events.

Reviewer: Is there any situation when it is triggered by the PR but **wouldn't** have already been triggered by the latest commit to the branch the PR was based on? I'm just wondering whether the distinction has an impact.

I probably should stop thinking about this and just make it say:
...'when changes are pushed to the branch *and* when a PR is created.' ?

Author: I can cite two situations, where
- the repo was forked and pr was created on the forked repo and to be merged on this repo. 
- there were multiple local commits and the latest commit wasn't on `paths` listed in the workflow, so the workflow wouldn't be triggered when all the commits were pushed on the remote branch.

- The integration test will be run on PR, merge and release events.

> - Once the PR is merged, a build will run. The resulted image will be tagged with git `head` and `latest` and pushed to ECR.

Reviewer: ```suggestion
- Once the PR is merged, a build will run. The resultant image will be tagged with git `head` and `latest` and pushed to ECR.
```

> - When decide to release a new stable image, create a release with a version number with `major.minor.patch`, e.g.,  `2.0.1`.

Reviewer: ```suggestion
- When you'd like to release a new stable image, create a release with a version number in the format `major.minor.patch`, e.g., `2.0.1`. This will trigger a build and the resultant image will be tagged with `2.0.1` (in this example) and `stable`, and pushed to ECR.
```

> It will trigger the build and image will be tagged with `2.0.1` and `stable` and pushed to ECR.

Reviewer: ```suggestion
```

## Packages' version maintenance and upgrade

> The base environment is using conda and docker image is built in two stages:

Reviewer: ```suggestion
The base environment uses Conda, and the Docker image is built in two stages:
```

Author: The base env is not the subject of "build". It's hard to define a subject in the context.

Reviewer: I feel like that can still apply to a system. Eg x _leverages_ y to perform function z.

Though I'll put it to the vote
@Ariana-B

Reviewer 2: If the base environment is not the one that builds the Docker image, then better to say 'and the docker image is built..." But I'm not really sure what this sentence is trying to say.

Reviewer: Hmm. Just realising I'm also confused. Base environment inside the first stage docker build... ?

Author: While I assume you all know how to build a docker image, and I seriously doubt if more words will help to clarify anything, I have no better way but sound like giving a  lesson on " what is a docker image build", and I apologise.

- For this sandbox image, conda is used as the base environment for further customised installation, but not included in the final image. So the base environment is only a starting point not even a component, and it does NOT perform any action. As for who or what is the subject for "build the docker image", or whether there should be a definitive subject, I guess it's not the real question here.
- multi-stage builds, ref https://docs.docker.com/build/building/multi-stage/

Less words, not sure if they'll cause confusion again though:
- start from conda
- two stage builds

Reviewer: @emmaai sorry I got too focussed on the phrasing, and not focussed enough on making sure I actually understood the [process](https://github.com/GeoscienceAustralia/dea-sandbox/blob/223276e1c5dd6665ff52a6180b98ea3caba68c50/docker/Dockerfile#LL27C37-L27C44). Thanks for that Docker doco as a refresher.

> 1. Create conda env and install as many as possible packages from `conda-forge`. Then `pip install` the rest, e.g., most `odc-` packages.

Reviewer: ```suggestion
1. Creates a Conda env and installs as many of the required packages as possible from `conda-forge`. It then runs `pip install` to install the rest, e.g., most `odc-` packages.
```

> 2. Copy the conda env to a new ubuntu image.

Reviewer: ```suggestion
2. Copies the Conda env to a new Ubuntu image.
```

> To speed up the building, the build workflow pulls the cache stored on ECR. Though with every build, the cache layers starting from `pip install` will be busted such that the newest versions of `odc-` packages will be installed. Thus, for version upgrade on these packages, creating a release will be enough.

Reviewer: 'busted'? As in they'll be wrong, or they'll get evicted from the cache first?
```suggestion
To speed up the build, the workflow pulls images from a cache stored on ECR. However, with every build the cache layers starting from `pip install` will be discarded, so that the newest versions of `odc-` packages will be installed. Thus, to perform version upgrades on these packages, creating a release is sufficient.
```

Author: They'll be deemed wrong (disregarded, ignored) in the building process. What happens to the cached image depends on other factors, not a concern here.

> The old conda env cache is used for all builds unless `env.yml` is changed. Other than the reason of image building time cited above, it takes more effort to resolve the conflicts and maintain the code base of `odc-` when the major upgrade on some base packages happens, e.g., some geospatial related packages, `GDAL` and `GEOS`. Hence, after evaluate pros and cons, we decide to keep this part rather manual.

Reviewer: I'm not sure whether I butchered this one.
```suggestion
The old Conda env cache is used for all builds unless `env.yml` is changed. In addition to speeding up builds, this cached environment allows us to maintain a working `odc-` codebase and defer resolving conflicts on geospatial base packages such as `GDAL` and `GEOS`, until we have a good opportunity to manually review them.
```

Author: I've observed some fun linguistic facts here. I use context aware expression subconsciously. "Other than..." means "besides, in addition to" within the context, but the full sentence is "I don't really have a good excuse of using conda cache other than the time reason mentioned above, but I do prefer solving the conflicts less frequently since a) it causes trouble and requires effort when conflicts happens, b)it doesn't seem necessary so far". And whether it's ideal or the best compromise is inconclusive, it's only a decision based on the current preference. I guess the only thing I'm trying to emphasise here is that the upgrade of conda env is manual and preferable atm.

Reviewer: Ahh 😄
And for whether or not the option was the best compromise, I think I read it as more conclusive from the `after evaluate pros and cons, we decide` bit.

> The steps for package versions upgrade are as follows:

Reviewer: ```suggestion
The steps for package version upgrades are as follows:
```

- For `odc-` packages, create a new release.
- For packages listed in `env.yml`, be specific with the version required, for example, `Shapely>=2.0`.

> **Note**: Avoid using `==` or `<=`, unless there is a hard requirement or a very good reason. The reason must be very specific in your PR, such as required from what, on what base and the linked PR (if applicable). 

Reviewer: ```suggestion
**Note**: Avoid using `==` or `<=`, unless there is a hard requirement or a very good reason. Ensure you specify this reason clearly in your PR; what is it required for, your justification, and any supporting PRs (if applicable).
```
